### PR TITLE
feat: Add validation for Hyper-V on Windows

### LINKED ISSUE #46: Add validation for Hyper-V on Windows
This is required to get fast performing Android emulators.

Validate that either HAXM or Hyper-V is installed (https://github.com/unoplatform/uno.check/issues/46#issuecomment-1099614653)

### PR DESCRIPTION
**Pull request by aesteves900**
GitHub Issue (If applicable): closes #46 

# PR Type

What kind of change does this PR introduce?

- Feature

## What is the current behavior?

People from the community have reported that certain projects fail to build because of long paths.

There is no check to Hyper-V activation on Windows.

## What is the new behavior?

The uno-check will validate the presence of Hyper-V and install if required.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current Figma desktop app
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/nventive/Uno/blob/master/doc/.feature-template.md). (for bug fixes / features)
- [x] Unit Tests for the changes have been added (for bug fixes / features) (if applicable)
- [x] [Wasm UI Tests](doc/articles/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes and the __PlugIn is BACKWARD COMPATIBLE__
- [x] Associated with an issue

### CODE REVIEW

File: UnoCheck/Solutions/HyperVActivationSolution.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+		}
+
+		private async Task<bool> ActivateHyperVAsync()
+		{
+			var startInfo = new ProcessStartInfo
+			{
+				Arguments = "/set hypervisorlaunchtype auto",
+				CreateNoWindow = true,
+				FileName = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.System), "bcdedit.exe"),
+				UseShellExecute = false
```
Comment (jeromelaban) [inline UnoCheck/Solutions/HyperVActivationSolution.cs:36]: I think we should only point to Microsoft's documentation, there are multiple machine specific steps that may need to be taken. The other validations are useful though, thank you.

File: UnoCheck/Checkups/HyperVCheckup.cs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+		public override async Task<DiagnosticResult> Examine(SharedState history)
+		{
+			
+			if (!await this.GetHyperVStatusAsync())
+			{
+				return await Task.FromResult(new DiagnosticResult(
+				Status.Error,
+				this,
+				new Suggestion("Activate HyperV on this computer? The computer must be BIOS virtualization enabled first!",
```
Comment (jeromelaban) [inline UnoCheck/Checkups/HyperVCheckup.cs:28]:
```suggestion
				new Suggestion("Activate HyperV on this computer to benefit from faster Android Emulators.",
```

File: UnoCheck/Checkups/HyperVCheckup.cs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+				new HyperVActivationSolution())));
+			}
+			else
+			{
+				var r = ShellProcessRunner.Run("shutdown", "/a");
+
+				ReportStatus(r.GetOutput(), Status.Error);
+
+
+				ReportStatus($"Hyper-V is configured!", Status.Ok);
```
Comment (jeromelaban) [inline UnoCheck/Checkups/HyperVCheckup.cs:38]: Let's just report the need for the computer to be restarted, or point to the documentation.

**Review by jeromelaban — APPROVED** [review]
Looking good!

File: UnoCheck/Native/Windows/SafeNativeMethods.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
+﻿using System;
```
Comment (ebariche) [inline UnoCheck/Native/Windows/SafeNativeMethods.cs:1]: Where is this file used ?

File: UnoCheck/Solutions/HyperVActivationSolution.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+{
+	public class HyperVActivationSolution : Solution
+	{
+		public HyperVActivationSolution()
+		{
+		}
+
+		public override async Task Implement(SharedState sharedState, CancellationToken cancellationToken)
+		{
+			await base.Implement(sharedState, cancellationToken);
```
Comment (ebariche) [inline UnoCheck/Solutions/HyperVActivationSolution.cs:17]: No need to call the base class.

File: UnoCheck/Solutions/HyperVActivationSolution.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+			if (await this.ActivateHyperVAsync())
+			{
+				ReportStatus("Activating Hyper-V. Please, restart computer to complete HyperV activation process.");
```
Comment (ebariche) [inline UnoCheck/Solutions/HyperVActivationSolution.cs:21]: restart your computer

Comment (ebariche) [reply UnoCheck/Solutions/HyperVActivationSolution.cs:21]: the Hyper-V activation

File: UnoCheck/Solutions/HyperVActivationSolution.cs [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+			}
+			else
+			{
+				throw new InvalidOperationException("Please, restart computer to complete HyperV activation process.");
```
Comment (ebariche) [inline UnoCheck/Solutions/HyperVActivationSolution.cs:25]: restart your computer

Comment (ebariche) [reply UnoCheck/Solutions/HyperVActivationSolution.cs:25]: the Hyper-V activation

File: UnoCheck/Solutions/HyperVActivationSolution.cs [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+			}
+		}
+
+		private async Task<bool> ActivateHyperVAsync()
+		{
+			string system32Directory = Path.Combine(Environment.ExpandEnvironmentVariables("%windir%"), "system32", "dism.exe");
```
Comment (ebariche) [inline UnoCheck/Solutions/HyperVActivationSolution.cs:31]: system32Directory -> dismPath

File: UnoCheck/Solutions/HyperVActivationSolution.cs [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+			{
+				ReportStatus("Activating Hyper-V. Please, restart computer to complete HyperV activation process.");
+			}
+			else
+			{
+				throw new InvalidOperationException("Please, restart computer to complete HyperV activation process.");
+			}
+		}
+
+		private async Task<bool> ActivateHyperVAsync()
```
Comment (ebariche) [inline UnoCheck/Solutions/HyperVActivationSolution.cs:29]: Why is this async ?

Comment (aesteves900) [reply UnoCheck/Solutions/HyperVActivationSolution.cs:27]: I Removed the async.

File: UnoCheck/Checkups/HyperVCheckup.cs [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+using System.IO;
+using System.Threading.Tasks;
+
+namespace DotNetCheck.Checkups
+{
+	public class HyperVCheckup : Checkup
+	{
+		public override string Id => "windowshyperv";
+
+		public override string Title => $"Windows {FeatureName} Checkup";
```
Comment (ebariche) [inline UnoCheck/Checkups/HyperVCheckup.cs:16]: Why not a search and replace ?

File: UnoCheck/Checkups/HyperVCheckup.cs [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+		public override bool IsPlatformSupported(Platform platform) => platform == Platform.Windows;
+
+		public override Task<DiagnosticResult> Examine(SharedState history)
+		{
+
```
Comment (ebariche) [inline UnoCheck/Checkups/HyperVCheckup.cs:22]: Remove empty line

File: UnoCheck/Native/Windows/SafeNativeMethods.cs (hunk 4 above)
Comment (ebariche) [reply UnoCheck/Native/Windows/SafeNativeMethods.cs:1]: ?

File: UnoCheck/Checkups/HyperVCheckup.cs [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+				CreateNoWindow = true,
+				FileName = Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.System), "bcdedit.exe"),
+				RedirectStandardOutput = true,
+				UseShellExecute = false
+			};
+			using (var process = Process.Start(startInfo))
+			{
+				process.WaitForExit();
+
+				if (process is not null)
```
Comment (ebariche) [inline UnoCheck/Checkups/HyperVCheckup.cs:53]: Not possible

File: UnoCheck/Checkups/HyperVCheckup.cs [hunk 13]
```diff
@@ -0,0 +1,74 @@
...
+				{
+					while (!process.StandardOutput.EndOfStream)
+					{
+						string? line = process?.StandardOutput?.ReadLine();
```
Comment (ebariche) [inline UnoCheck/Checkups/HyperVCheckup.cs:57]: Those null checks are not necessary !

File: UnoCheck/Solutions/HyperVActivationSolution.cs (hunk 9 above)
Comment (ebariche) [reply UnoCheck/Solutions/HyperVActivationSolution.cs:27]: It's still there !

File: UnoCheck/Solutions/HyperVActivationSolution.cs [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+	{
+		public HyperVActivationSolution()
+		{
+		}
+
+		public override async Task Implement(SharedState sharedState, CancellationToken cancellationToken)
+		{
+			if (await this.ActivateHyperVAsync())
+			{
+				ReportStatus("The Hyper-V activated. Please, restart your computer.");
```
Comment (ebariche) [inline UnoCheck/Solutions/HyperVActivationSolution.cs:19]: "Hyper-V activated" not "The Hyper-V activated"

File: UnoCheck/Solutions/HyperVActivationSolution.cs [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+				// For 32-bit processes on 64-bit systems, %windir%\system32 folder
+				// can only be accessed by specifying %windir%\sysnative folder.
+				dsimPath = Path.Combine(Environment.ExpandEnvironmentVariables("%windir%"), "sysnative", "dism.exe");
+			}
+
+			var dism = ShellProcessRunner.Run(dsimPath, "/Online /Enable-Feature /Quiet /NoRestart /All /FeatureName:Microsoft-Hyper-V");
+
+			if(dism.ExitCode == RebootRequired)
+			{
+				ReportStatus("Please, restart computer to complete the Hyper-V activation process.");
```
Comment (ebariche) [inline UnoCheck/Solutions/HyperVActivationSolution.cs:42]: restart your computer

File: UnoCheck/Checkups/HyperVCheckup.cs [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+		public override bool IsPlatformSupported(Platform platform) => platform == Platform.Windows;
+
+		public override Task<DiagnosticResult> Examine(SharedState history)
+		{
+			if (!this.GetHyperVStatus())
+			{
+				return Task.FromResult(new DiagnosticResult(
+				Status.Warning,
+				this,
+				new Suggestion($"Activate Hyper-V on this computer to benefit from faster Android Emulators!",
```
Comment (jeromelaban) [inline UnoCheck/Checkups/HyperVCheckup.cs:27]:
```suggestion
				new Suggestion($"Activate Hyper-V on this computer to benefit from faster Android Emulators",
```

File: UnoCheck/Checkups/HyperVCheckup.cs [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+			{
+				return Task.FromResult(new DiagnosticResult(
+				Status.Warning,
+				this,
+				new Suggestion($"Activate Hyper-V on this computer to benefit from faster Android Emulators",
+				new HyperVActivationSolution())));
+			}
+			else
+			{
+				ReportStatus($"Hyper-V is configured!", Status.Ok);
```
Comment (jeromelaban) [inline UnoCheck/Checkups/HyperVCheckup.cs:32]:
```suggestion
				ReportStatus($"Hyper-V is configured", Status.Ok);
```

**Review by jeromelaban — APPROVED** [review]

**Pull request merged**